### PR TITLE
Preliminary command-line options parser

### PR DESCRIPTION
**Pull request by pelesh**
Provides:
- Basic parser for command line options.
- Single test for iterative solvers through SystemSolver interface

### CODE REVIEW

**Comment by pelesh** [comment]
@kswirydo, after rebasing to the latest `develop`, unit tests for CGS1 pass, but functionality tests using CGS1 still fail, bot on CPU and GPU.

File: resolve/CMakeLists.txt [hunk 1]
```diff
@@ -127,7 +127,7 @@ target_include_directories(ReSolve INTERFACE
 )
 
 target_link_libraries(ReSolve PUBLIC ${ReSolve_Targets_List})
-target_link_libraries(ReSolve PRIVATE resolve_version)
+target_link_libraries(ReSolve PRIVATE resolve_version resolve_options)
```
Comment (cameronrutherford) [inline resolve/CMakeLists.txt:130]: I was going to PR a quick fix to this but got distracted/ran out of time... I just wonder if this library (and other libraries like `resolve_version` should be added instead just to `${ReSolve_Targets_List}`, or if we should leave this logic to the relevant `CMakeLists.txt` and try and clean up this core `resolve/CMakeLists.txt` a little.

My suggestion here is purely cosmetic though...

File: resolve/LinSolverIterativeRandFGMRES.cpp [hunk 2]
```diff
@@ -430,6 +430,21 @@ namespace ReSolve
     return 0;
   }
 
+  int LinSolverIterativeRandFGMRES::setSketchingMethod(std::string method)
+  {
+    if (method == "count") {
```
Comment (cameronrutherford) [inline resolve/LinSolverIterativeRandFGMRES.cpp:435]: Similar with the magic strings comment somewhere else, might be nice to standardize how these options are handled.

File: resolve/SystemSolver.cpp [hunk 3]
```diff
@@ -281,12 +289,18 @@ namespace ReSolve
                        << "Using default.\n";
         sketch = LinSolverIterativeRandFGMRES::cs;
       }
-
       setGramSchmidtMethod(gsMethod_);
       iterativeSolver_ = new LinSolverIterativeRandFGMRES(matrixHandler_,
                                                           vectorHandler_,
                                                           sketch,
                                                           gs_);
+    } else if (solveMethod_ == "fgmres") {
```
Comment (cameronrutherford) [inline resolve/SystemSolver.cpp:297]: Another magic string...

File: resolve/SystemSolver.cpp [hunk 4]
```diff
@@ -419,7 +433,7 @@ namespace ReSolve
     int status = 0;
 
     // Use Krylov solver if selected
-    if (solveMethod_ == "randgmres") {
+    if (solveMethod_ == "randgmres" || solveMethod_ == "fgmres") {
```
Comment (cameronrutherford) [inline resolve/SystemSolver.cpp:436]: More magic strings

File: resolve/utilities/params/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+add_library(resolve_options OBJECT ${Options_SRC})
+set_property(TARGET resolve_options PROPERTY POSITION_INDEPENDENT_CODE ON)
+
+target_include_directories(resolve_options PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>
+    $<BUILD_INTERFACE:${CMAKE_BINARY_DIR}>
+    $<INSTALL_INTERFACE:include>
+)
+
+install(FILES ${Options_HEADER_INSTALL} DESTINATION include/resolve/utilities/options)
```
Comment (cameronrutherford) [inline resolve/utilities/params/CMakeLists.txt:27]:
```suggestion
install(FILES ${Options_HEADER_INSTALL} DESTINATION include/resolve/utilities/options)

target_link_libraries(...)
```

Per my previous comments, we would just be moving where the `target_link_libraries` call actually happens. Again, purely a cosmetic suggestion

File: resolve/utilities/params/CliOptions.cpp [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+  const char* const *CliOptions::end() const
+  {
+      return argv_ + argc_;
+  }
+
+  const char* const *CliOptions::last() const
+  {
+      return argv_ + argc_ - 1;
+  }
```
Comment (cameronrutherford) [inline resolve/utilities/params/CliOptions.cpp:127]: Perhaps some documentation about options would be nice so we at least have a consistent user experience :)

Also, do we want to handle options from a file as well?

File: tests/functionality/CMakeLists.txt [hunk 7]
```diff
@@ -121,24 +107,39 @@ if(RESOLVE_USE_KLU)
...
-if(NOT RESOLVE_USE_GPU)
-  add_test(NAME sys_rand_gmres_cpu_test COMMAND $<TARGET_FILE:rand_gmres_cpu_test.exe>)
-endif()
+# Krylov solvers tests
+#add_test(NAME sys_rand_count_gmres_cgs1_test COMMAND $<TARGET_FILE:sys_rand_gmres_test.exe>     "-i" "randgmres" "-g" "cgs1" "-s" "count")
+add_test(NAME sys_rand_count_gmres_cgs2_test COMMAND $<TARGET_FILE:sys_rand_gmres_test.exe>     "-i" "randgmres" "-g" "cgs2" "-s" "count")
+add_test(NAME sys_rand_count_gmres_mgs_test COMMAND $<TARGET_FILE:sys_rand_gmres_test.exe>      "-i" "randgmres" "-g" "mgs" "-s" "count")
+add_test(NAME sys_rand_count_gmres_mgs2sync_test COMMAND $<TARGET_FILE:sys_rand_gmres_test.exe> "-i" "randgmres" "-g" "mgs_two_synch" "-s" "count")
+add_test(NAME sys_rand_count_gmres_mgspm_test COMMAND $<TARGET_FILE:sys_rand_gmres_test.exe>    "-i" "randgmres" "-g" "mgs_pm" "-s" "count")
+#add_test(NAME sys_rand_fwht_gmres_cgs1_test COMMAND $<TARGET_FILE:sys_rand_gmres_test.exe>      "-i" "randgmres" "-g" "cgs1" "-s" "fwht")
```
Comment (cameronrutherford) [inline tests/functionality/CMakeLists.txt:116]: Do we plan on adding these commented out tests back in later?

File: tests/functionality/testSysRandGMRES.cpp [hunk 8]
```diff
@@ -55,8 +62,23 @@ int main(int argc, char *argv[])
...
+  const index_type N = opt ? atoi((*opt).second.c_str()) : 10000;
+
+  opt = options.getParamFromKey("-i");
+  std::string method = opt ? (*opt).second : "randgmres";
+
+  opt = options.getParamFromKey("-g");
+  std::string gs = opt ? (*opt).second : "cgs2";
+
+  opt = options.getParamFromKey("-s");
+  std::string sketch = opt ? (*opt).second : "count";
```
Comment (cameronrutherford) [inline tests/functionality/testSysRandGMRES.cpp:79]: I am not thrilled about this implementation, but it does work well. Just need to make sure that defaults are standardized, and hopefully ensure we copy and paste the right snippets when making different tests

File: resolve/SystemSolver.cpp [hunk 9]
```diff
@@ -261,7 +268,8 @@ namespace ReSolve
 #elif defined(RESOLVE_USE_HIP)
       preconditioner_ = new LinSolverDirectRocSparseILU0(workspaceHip_);
 #else
-      preconditioner_ = new LinSolverDirectSerialILU0(workspaceCpu_);
+      // preconditioner_ = new LinSolverDirectSerialILU0(workspaceCpu_);
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/SystemSolver.cpp:271]: We have two alternate implementations of ILU0 for now and will decide about it later (probably pick best features from each). That however is outside the scope of this PR.

File: resolve/SystemSolver.cpp [hunk 10]
```diff
@@ -690,15 +714,32 @@ namespace ReSolve
...
 
   /**
-   * @brief Selekt sketching method for randomized solvers
+   * @brief Select sketching method for randomized solvers
+   * 
+   * This is a brute force method that will delete randomized GMRES solver
+   * only to change its sketching function.
    * 
-   * This needs to be moved to LinSolverIterative class and accessed from there.
+   * @todo This needs to be moved to LinSolverIterative class and accessed from there.
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/SystemSolver.cpp:722]: See #96. You are tagged there :)

File: resolve/SystemSolver.cpp [hunk 11]
```diff
@@ -690,15 +714,32 @@ namespace ReSolve
...
    * 
-   * @param sketching_method 
+   * @param[in] sketching_method - string ID of the sketching method
    */
-  void SystemSolver::setSketchingMethod(std::string sketching_method)
+  int SystemSolver::setSketchingMethod(std::string sketching_method)
   {
-    sketching_method_ = sketching_method;
+    if (solveMethod_ != "randgmres") {
```
[the remark this answers is not in the document]
Comment (pelesh) [reply resolve/SystemSolver.cpp:728]: These strings will come from CLI. You are right that internally they need to be handled by enums, but this is again something we need to discuss further in #96.

File: resolve/SystemSolver.cpp (hunk 4 above)
Comment (pelesh) [reply resolve/SystemSolver.cpp:436]: That is a good point. See #96.

File: resolve/utilities/params/CliOptions.cpp (hunk 6 above)
Comment (pelesh) [reply resolve/utilities/params/CliOptions.cpp:127]: `CliOPtions` is just parsing whatever is on command line input. It does not specify any parameters nor it is intended to do so. Parameter management is something we still need to design and implement. It is tracked in #96 (and I guess you can see a pattern in my responses here).

File: resolve/utilities/params/CMakeLists.txt (hunk 5 above)
Comment (pelesh) [reply resolve/utilities/params/CMakeLists.txt:27]: We still need two different calls for linking libraries. One is for public linking of shared libraries and the other one is for private linking of object libraries.

File: tests/functionality/testSysRandGMRES.cpp (hunk 8 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp:79]: This will come from the parameter management system that we are yet to implement. The value of this make shift solution is that we can now test every single configuration of every iterative solver we implemented so far from a single test binary. This will also help standardize interfaces as we implement other iterative solvers.

File: resolve/CMakeLists.txt (hunk 1 above)
Comment (pelesh) [reply resolve/CMakeLists.txt:130]: I added a separate `ReSolve_Object_List`. We should perhaps consider building more components of ReSolve as object rather than shared libraries.

File: resolve/utilities/params/CMakeLists.txt (hunk 5 above)
Comment (pelesh) [reply resolve/utilities/params/CMakeLists.txt:27]: I think we should refactor CMake per your suggestion. That would make configuration easier to understand. See #145

File: tests/functionality/CMakeLists.txt (hunk 7 above)
Comment (pelesh) [reply tests/functionality/CMakeLists.txt]: Addressed in [e3b05b5](https://github.com/ORNL/ReSolve/pull/140/commits/e3b05b54b24e278858ba1f95a9b74b5391b37a3b)

File: resolve/CMakeLists.txt (hunk 1 above)
Comment (pelesh) [reply resolve/CMakeLists.txt]: See [e3b05b5](https://github.com/ORNL/ReSolve/pull/140/commits/e3b05b54b24e278858ba1f95a9b74b5391b37a3b)

**Review by cameronrutherford — APPROVED** [review]

File: tests/functionality/testSysRandGMRES.cpp [hunk 12]
```diff
@@ -71,116 +93,139 @@ int main(int argc, char *argv[])
...
 
   // Create system solver
-  ReSolve::SystemSolver* solver = new ReSolve::SystemSolver(&workspace, "none", "none", "randgmres", "ilu0", "none");
+  ReSolve::SystemSolver solver(&workspace, "none", "none", method, "ilu0", "none");
+  solver.setGramSchmidtMethod(gs);
 
   // Create solution vector
-  vector_type* vec_x = new vector_type(A->getNumRows());
-  vec_x->allocate(ReSolve::memory::HOST);
+  vector_type vec_x(A->getNumRows());
```
Comment (kswirydo) [inline tests/functionality/testSysRandGMRES.cpp:100]: Not sure what is the reason for this change

File: tests/functionality/testSysRandGMRES.cpp [hunk 13]
```diff
@@ -71,116 +93,139 @@ int main(int argc, char *argv[])
...
+  status = solver.solve(vec_rhs, &vec_x);
   error_sum += status;
 
   // Get residual norm
   norm_b = vector_handler.dot(vec_rhs, vec_rhs, memspace);
   norm_b = std::sqrt(norm_b);
-  real_type final_norm_first =  solver->getIterativeSolver().getFinalResidualNorm();
-  std::cout << std::scientific << std::setprecision(16)
-            << "Randomized FGMRES results (first run): \n"
-            << "\t Sketching method:                                    : "
```
Comment (kswirydo) [inline tests/functionality/testSysRandGMRES.cpp:112]: Can you please put this info back? It is helpful.

Comment (kswirydo) [reply tests/functionality/testSysRandGMRES.cpp:112]: Unless you display it in header info...

**Review by kswirydo — APPROVED** [review]
Looks good. Now I am running it to see how the CLO interface works in practice.

File: tests/functionality/testSysRandGMRES.cpp (hunk 12 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp:100]: Just opting to create vectors in the local scope to avoid using pointers.

File: tests/functionality/testSysRandGMRES.cpp (hunk 13 above)
Comment (pelesh) [reply tests/functionality/testSysRandGMRES.cpp:112]: You are correct, they are printed by `headerInfo` function, which processes CLI parameters and prints human readable info about solver configuration.

**Comment by kswirydo** [comment]
Also this (note there is no sketching method so it is set to default - but then the name is not displayed)
```
./tests/functionality/sys_rand_gmres_test.exe -i randgmres  -g cgs3
Sketching method  not recognized.
Setting sketching to the default (count).

[WARNING] Gram-Schmidt variant cgs2 not recognized.
[WARNING] Using default cgs2 Gram-Schmidt variant.
[WARNING] Sketching method  not recognized!
Using default.
Iterative solver results: 
	 Sketching method:                                    : 
	 Initial residual norm:          ||b-Ax_0||_2         : 7.8910952345032565e+03 
	 Initial relative residual norm: ||b-Ax_0||_2/||b||_2 : 1.0053370321875226e+00 
	 Final residual norm:            ||b-Ax||_2           : 1.1069282838889689e-08 
	 Final relative residual norm:   ||b-Ax||_2/||b||_2   : 1.4102427644055879e-12 
	 Number of iterations                                 : 19
Test PASSED
```

**Comment by pelesh** [comment]
If the option parameter is unrecognized, the test should fall back to the default. It should not segfault. I agree, sketching should not be printed, if randomized method is not used. This is easy to fix though.

**Comment by pelesh** [comment]
> Also this (note there is no sketching method so it is set to default - but then the name is not displayed)
> 
> ```
> ./tests/functionality/sys_rand_gmres_test.exe -i randgmres  -g cgs3
> Sketching method  not recognized.
> Setting sketching to the default (count).
> 
> [WARNING] Gram-Schmidt variant cgs2 not recognized.
> [WARNING] Using default cgs2 Gram-Schmidt variant.
> [WARNING] Sketching method  not recognized!
> Using default.
> Iterative solver results: 
> 	 Sketching method:                                    : 
> 	 Initial residual norm:          ||b-Ax_0||_2         : 7.8910952345032565e+03 
> 	 Initial relative residual norm: ||b-Ax_0||_2/||b||_2 : 1.0053370321875226e+00 
> 	 Final residual norm:            ||b-Ax||_2           : 1.1069282838889689e-08 
> 	 Final relative residual norm:   ||b-Ax||_2/||b||_2   : 1.4102427644055879e-12 
> 	 Number of iterations                                 : 19
> Test PASSED
> ```

I cannot reproduce this. I get this when I make the same call:
```
$ ./tests/functionality/sys_rand_gmres_test.exe -i randgmres  -g cgs3
Unknown orthogonalization cgs3
Setting orthogonalization to the default (CGS2).

Results for randomized FGMRES solver
	 Sketching method:               count sketching
	 Orthogonalization method:       reorthogonalized classical Gram-Schmidt
	 Hardware backend:               HIP
	 Solver tolerance:               1e-12
	 Initial residual norm:          ||b-Ax_0||_2         : 7.9193852034106785e+03 
	 Initial relative residual norm: ||b-Ax_0||_2/||b||_2 : 1.0089412154519328e+00 
	 Final residual norm:            ||b-Ax||_2           : 5.0631784499176610e-09 
	 Final relative residual norm:   ||b-Ax||_2/||b||_2   : 6.4505631284482523e-13 
	 Number of iterations                                 : 19
Test PASSED
```
I pulled from the lates `cli-params-dev`.

**Comment by pelesh** [comment]
@kswirydo, I couldn't reproduce issues you reported. I will merge this but feel free to open an issue, if you still see these problems.
